okhttp/src/main/java/okhttp3/HttpUrl.java

```diff
@@ -945,6 +947,50 @@ public static HttpUrl get(URI uri) {
     return url;
   }
 
+  /**
+   * Returns the domain name of this URL's {@link #host()} that is one level beneath the public
+   * suffix by consulting the <a href="https://publicsuffix.org">public suffix list</a>. Returns
+   * null if this URL's {@link #host()} is an IP address or is considered a public suffix by the
+   * public suffix list.
+   *
+   * <p>Use {@link #isPublicSuffix()} to differentiate between private hosts and routable
+   * non-private hosts that might be public suffixes.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>{@code topPrivateDomain()}</th></tr>
+   *   <tr><td>{@code http://google.com}</td><td>{@code "google.com"}</td></tr>
+   *   <tr><td>{@code http://adwords.google.co.uk}</td><td>{@code "google.co.uk"}</td></tr>
+   *   <tr><td>{@code http://square}</td><td>null</td></tr>
+   *   <tr><td>{@code http://co.uk}</td><td>null</td></tr>
+   *   <tr><td>{@code http://localhost}</td><td>null</td></tr>
+   *   <tr><td>{@code http://127.0.0.1}</td><td>null</td></tr>
+   * </table>
+   */
+  public String topPrivateDomain() {
+    if (verifyAsIpAddress(host)) return null;
+    return PublicSuffixDatabase.get().getEffectiveTldPlusOne(host);
+  }
+
+  /**
+   * Returns true if this URL's {@link #host()} is a public suffix contained in the
+   * <a href="https://publicsuffix.org">public suffix list</a>. This will return false if the
+   * {@link #host()} does not explicitly appear in the public suffix list.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>{@code isPublicSuffix()}</th></tr>
+   *   <tr><td>{@code http://com}</td><td>{@code true}</td></tr>
+   *   <tr><td>{@code http://co.uk}</td><td>{@code true}</td></tr>
+   *   <tr><td>{@code http://square}</td><td>{@code false}</td></tr>
+   *   <tr><td>{@code http://google.com}</td><td>{@code false}</td></tr>
+   *   <tr><td>{@code http://localhost}</td><td>{@code false}</td></tr>
+   *   <tr><td>{@code http://127.0.0.1}</td><td>{@code false}</td></tr>
+   * </table>
+   */
+  public boolean isPublicSuffix() {
```

Comment:
(and if I’m missing something, I still think it might want to be a static method like `HttpUrl.isPublicSuffix(String)`. Unsure though.